src/wallet/wallet.cpp

```diff
@@ -1140,22 +1140,23 @@ bool CWallet::LoadCaches()
     // the default address for each key was automatically added to the in-memory
     // keystore, but not persisted. Following the addition of unified addresses,
     // all addresses must be written to the wallet database explicitly.
-    auto legacySeed = GetLegacyHDSeed();
-    if (legacySeed.has_value()) {
-        for (const auto& [saplingIvk, keyMeta] : mapSaplingZKeyMetadata) {
-            // This condition only applies for keys derived from the legacy seed
-            if (keyMeta.seedFp == legacySeed.value().Fingerprint()) {
-                SaplingExtendedFullViewingKey extfvk;
-                if (GetSaplingFullViewingKey(saplingIvk, extfvk)) {
-                    auto defaultAddress = extfvk.DefaultAddress();
-                    if (!HaveSaplingIncomingViewingKey(defaultAddress)) {
-                        // restore the address to the keystore and persist it so that
-                        // the database state is consistent.
-                        if (!AddSaplingPaymentAddress(saplingIvk, defaultAddress)) {
-                            LogPrintf("%s: Error: Failed to write legacy Sapling payment address to the wallet database.\n",
-                                __func__);
-                            return false;
-                        }
+    auto mnemonicSeed = GetMnemonicSeed();
+    for (const auto& [saplingIvk, keyMeta] : mapSaplingZKeyMetadata) {
+        // This condition only applies for keys derived from the legacy seed
+        // or from imported keys.
+        if (!mnemonicSeed.has_value() || keyMeta.seedFp != mnemonicSeed.value().Fingerprint()) {
```

Comment:
This is going to become incorrect if we add UFVK importing (as that will not necessarily be derived from the wallet's mnemonic), which would cause the default Sapling address to be added irrespective of any UA knowledge. Ensure we have this code referenced in the issue for adding UFVK importing, so we do not forget.